stan/math/prim/mat/prob/bernoulli_logit_glm_lpmf.hpp

```diff
@@ -81,63 +89,63 @@ typename return_type<T_x, T_alpha, T_beta>::type bernoulli_logit_glm_lpmf(
+
+  T_y_val signs = 2 * as_array_or_scalar(y_val_vec) - 1;
+
+  Eigen::Array<T_partials_return, Dynamic, 1> ytheta = x_val * beta_val_vec;
+  ytheta = as_array_or_scalar(signs)
+           * (ytheta + as_array_or_scalar(alpha_val_vec));
 
   // Compute the log-density and handle extreme values gracefully
   // using Taylor approximations.
   // And compute the derivatives wrt theta.
   static const double cutoff = 20.0;
-  Matrix<T_partials_return, Dynamic, 1> theta_derivative(N, 1);
-  T_partials_return theta_derivative_sum = 0;
-  T_partials_return exp_m_ythetan;
-  for (size_t n = 0; n < N; ++n) {
-    ytheta[n] += signs[n] * value_of(alpha_vec[n]);
-    exp_m_ythetan = exp(-ytheta[n]);
-    if (ytheta[n] > cutoff) {
-      logp -= exp_m_ythetan;
-      theta_derivative[n] = -exp_m_ythetan;
-    } else if (ytheta[n] < -cutoff) {
-      logp += ytheta[n];
-      theta_derivative[n] = signs[n];
-    } else {
-      logp -= log1p(exp_m_ythetan);
-      theta_derivative[n] = signs[n] * exp_m_ythetan / (exp_m_ythetan + 1);
+  Eigen::Array<T_partials_return, Dynamic, 1> exp_m_ytheta = exp(-ytheta);
+  logp += (ytheta > cutoff)
+              .select(-exp_m_ytheta,
+                      (ytheta < -cutoff).select(ytheta, -log1p(exp_m_ytheta)))
+              .sum();
+  if (!std::isfinite(logp)) {
+    check_finite(function, "Weight vector", beta);
+    check_finite(function, "Intercept", alpha);
+    for (size_t n = 0; n < N; ++n) {
+      check_finite(function, "Matrix of independent variables", ytheta[n]);
     }
-    if (!is_vector<T_alpha>::value)
-      theta_derivative_sum += theta_derivative[n];
   }
 
   // Compute the necessary derivatives.
   operands_and_partials<T_x, T_alpha, T_beta> ops_partials(x, alpha, beta);
-  if (!is_constant_struct<T_beta>::value) {
-    ops_partials.edge3_.partials_ = value_of(x).transpose() * theta_derivative;
-  }
-  if (!is_constant_struct<T_x>::value) {
-    ops_partials.edge1_.partials_ = theta_derivative * beta_dbl.transpose();
-  }
-  if (!is_constant_struct<T_alpha>::value) {
-    if (is_vector<T_alpha>::value)
-      ops_partials.edge2_.partials_ = theta_derivative;
-    else
-      ops_partials.edge2_.partials_[0] = theta_derivative_sum;
+  if (!is_constant_struct<T_beta>::value || !is_constant_struct<T_x>::value
+      || !is_constant_struct<T_alpha>::value) {
+    Matrix<T_partials_return, Dynamic, 1> theta_derivative
+        = (ytheta > cutoff)
+              .select(-exp_m_ytheta,
+                      (ytheta < -cutoff)
+                          .select(as_array_or_scalar(signs),
+                                  as_array_or_scalar(signs) * exp_m_ytheta
+                                      / (exp_m_ytheta + 1)));
+    if (!is_constant_struct<T_beta>::value) {
+      ops_partials.edge3_.partials_ = x_val.transpose() * theta_derivative;
+    }
+    if (!is_constant_struct<T_x>::value) {
+      ops_partials.edge1_.partials_
+          = (beta_val_vec * theta_derivative.transpose()).transpose();
```

Comment:
I think this is `(a * b')'`, which I think just works out to `b * a'`, i.e., `theta_derivative.transpose() * beta_val_vec`.

Reply:
This is intentional as beta_val_vec could be a scalar that does not define .transpose(). If you prefer more readability here at expense of some code elsewhere I can make a transpose() function that will be a no-op for scalars.

Reply:
I see.  It's OK as it is.  This is the kind of thing that deserves a comment, though, as it looks wrong at first glance without knowing that `beta_val_vec` can be a scalar.